File: Module2/DataParallelism.cs/Pipelines/Pipeline.cs
using System;
using System.Collections.Concurrent;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.FSharp.Core;
using static Pipeline.PipelineFunc;

namespace DataParallelism.Pipelines
{
    public class CsPipeline<TInput, TOutput> : IPipeline<TInput, TOutput>
    {
        private struct Continuation
        {
            public Func<Tuple<TInput, TOutput>, Unit> Callback;
            public TInput Input;
        }

        private readonly Func<TInput, TOutput> _function;
        private BlockingCollection<Continuation>[] _continuations;

        public CsPipeline(Func<TInput, TOutput> function)
        {
            _function = function;
        }

        // TODO : 2.4
        // (1)
        // implement a "Then" functions that compose the current _function with a given "next function"
        // assuming that the return type could be different from the current one "TOutput"
        // the return type is a new Pipeline, for example (conceptually), given a function "Func<TOutput, TMapped>",
        // the new Pipeline would have signature IPipeline<TInput, TMapped

        // ... CODE HERE
          public IPipeline<TInput, TMapped> Then<TMapped>(Func<TOutput, TMapped> nextfunction) => null;


        public void Enqueue(TInput input, Func<Tuple<TInput, TOutput>, Unit> callback)
        {
            BlockingCollection<Continuation>.TryAddToAny(_continuations,
                new Continuation
                {
                    Input = input,
                    Callback = callback
                });
        }

        public void Stop()
        {
            foreach (var bc in _continuations)
                bc.CompleteAdding();
        }

        public IDisposable Execute(int blockingCollectionPoolSize, CancellationToken cancellationToken)
        {
            _continuations =
                Enumerable.Range(0, blockingCollectionPoolSize)
                    .Select(_ => new BlockingCollection<Continuation>(100))
                    .ToArray();

            cancellationToken.Register(Stop);

            for (var x = 0; x < blockingCollectionPoolSize; x++)
                Task.Factory.StartNew(() =>
                {
                    while (!_continuations.All(bc => bc.IsCompleted) && !cancellationToken.IsCancellationRequested)
                    {
                        // TODO
                        // (2)
                        // implement a statement that runs the continuation from _continuations
                        // the Take continuation must be thread safe
                        // Check the BlockingCollection API
                        // Suggestion, use the Callback function to execute the continuation retrieved
                        // SUggestion, look into the BlockingCollection class API

                        Continuation continuation;

                    }
                }, cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Default);


            return new StopPipelineDisposable(this);
        }


        private class StopPipelineDisposable : IDisposable
        {
            readonly CsPipeline<TInput, TOutput> _pipeline;

            public StopPipelineDisposable(CsPipeline<TInput, TOutput> pipeline)
            {
                _pipeline = pipeline;
            }

            public void Dispose()
            {
                _pipeline.Stop();
            }
        }
    }
}

File: src/Module2/Pipeline.fs/Pipeline.fs
namespace Pipeline

open System
open System.Linq
open System.Collections.Concurrent
open System.Threading
open System.Threading.Tasks

//  The IPipeline interface
[<Interface>]
type IPipeline<'a,'b> =
    abstract member Then : ('b -> 'c) -> IPipeline<'a,'c>
    abstract member Enqueue : 'a * (('a * 'b) -> unit) -> unit
    abstract member Execute : int * CancellationToken -> IDisposable
    abstract member Stop : unit -> unit


// The Parallel Functional Pipeline pattern
[<Struct>]
type internal Continuation<'a, 'b>(input:'a, callback:('a * 'b) -> unit) =
    member this.Input with get() = input
    member this.Callback with get() = callback

type Pipeline<'a, 'b> private (func:'a -> 'b) as this =
    let continuations = Array.init 3 (fun _ -> new BlockingCollection<Continuation<'a,'b>>(100))

    // TODO : 2.16
    // implement a then' function that composes a given function to the current "func" one
    // Suggestion, use the ">>" operator

    // let then'  ...

    // #region solution
    let then' (nextFunction:'b -> 'c) =
        Pipeline(func >> nextFunction) :> IPipeline<_,_>
    // #endregion

    let enqueue (input:'a) (callback:('a * 'b) -> unit) =
        BlockingCollection<Continuation<_,_>>.AddToAny(continuations, Continuation(input, callback))

    let stop() = for continuation in continuations do continuation.CompleteAdding()

    let execute blockingCollectionPoolSize (cancellationToken:CancellationToken) =

        cancellationToken.Register(Action(stop)) |> ignore

        for i = 0 to blockingCollectionPoolSize - 1 do
            Task.Factory.StartNew(fun ( )->
                while (not <| continuations.All(fun bc -> bc.IsCompleted)) && (not <| cancellationToken.IsCancellationRequested) do
                    let continuation = ref Unchecked.defaultof<Continuation<_,_>>
                    BlockingCollection.TakeFromAny(continuations, continuation) |> ignore
                    let continuation = continuation.Value
                    continuation.Callback(continuation.Input, func(continuation.Input))

            ,cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Default) |> ignore

    static member Create(func:'a -> 'b) =
        Pipeline(func) :> IPipeline<_,_>

    interface IPipeline<'a, 'b> with
        // TODO : 2.16 remove "Unchecked.defaultof<IPipeline<_,_>>"  and apply the new function created, for example "then' nextFunction"
        member this.Then(nextFunction) = Unchecked.defaultof<IPipeline<_,_>> // then' nextFunction

        member this.Enqueue(input, callback) = enqueue input callback |> ignore
        member this.Stop() = stop()
        member this.Execute (blockingCollectionPoolSize, cancellationToken) =
            execute blockingCollectionPoolSize cancellationToken
            { new IDisposable with member self.Dispose() = stop() }
